kotlin/src/main/com/looker/rtl/SdkResult.kt

```diff
@@ -0,0 +1,104 @@
+package com.looker.rtl
+
+import io.ktor.client.call.receive
+import io.ktor.client.response.HttpResponse
+import io.ktor.http.isSuccess
+import kotlinx.coroutines.runBlocking
+
+class FailureResponseError(val result: SdkResult<*, *>) : Exception()
+
+interface SdkResponse<T> {
+    val success: Boolean
+    val statusCode: Int
+    val method: HttpMethod
+    val path: String
+    fun body(): T
+}
+
+sealed class SdkResult<out TSuccess, out TFailure> {
+    abstract val success: Boolean
+    abstract val method: HttpMethod
+    abstract val path: String
+
+    data class SuccessResponse<TSuccess>(
+        val response: HttpResponse,
+        override val method: HttpMethod,
+        override val path: String,
+        private val body: TSuccess
+    ) : SdkResponse<TSuccess>, SdkResult<TSuccess, Nothing>() {
+        override val success: Boolean = true
+        override val statusCode: Int = response.status.value
+
+        override fun body(): TSuccess = body
+
+        inline fun <reified T> bodyAs(): T {
+            return runBlocking { response.receive<T>() }
+        }
+    }
+
+    data class FailureResponse<TFailure>(
+        val response: HttpResponse,
+        override val method: HttpMethod,
+        override val path: String,
+        private val body: TFailure
+    ) : SdkResponse<TFailure>, SdkResult<Nothing, TFailure>() {
+        override val success: Boolean = false
+        override val statusCode: Int = response.status.value
+
+        override fun body(): TFailure = body
+
+        inline fun <reified T> bodyAs(): T {
+            return runBlocking { response.receive<T>() }
+        }
+    }
+
+    data class Error(
+        val error: Throwable,
+        override val method: HttpMethod,
+        override val path: String
+    ) : SdkResult<Nothing, Nothing>() {
+        override val success: Boolean = false
+    }
+
+    companion object {
+        inline fun <reified TSuccess, reified TFailure> response(
+            response: HttpResponse,
+            method: HttpMethod,
+            path: String
+        ): SdkResult<TSuccess, TFailure> {
+            try {
+                if (response.status.isSuccess()) {
+                    val body = runBlocking { response.receive<TSuccess>() }
+                    return SuccessResponse<TSuccess>(response, method, path, body)
+                } else {
+                    val body = runBlocking { response.receive<TFailure>() }
```

Comment:
Pondering this some more, I'm wondering if the FailureResponse case should actually defer on receiving the data until requested.  The issue I'm running into while thinking this through is that I want to be able to deal with a single API being able to return multiple different types of error response.  In that case, we're going to have to do some preliminary parsing or some try/catch(ParseError) for the different types.  With ktor, you can only do `response.receive()` a single time, so that's not going to work.

I'm thinking that maybe we do `response.receive<ByteArray>()` or `response.receive<String>()` to get the "raw" data, and then parse that into whatever type is requested.  Of course that'll mean that we have to manage our own object mapper in the RTL instead of using ktor's registered one.  All around, it ends up being a lot more to manage here in the RTL layer, but I'm not seeing a good way around that.

Looking at the error responses and some of the libraries like Jackson or kotlinx.serialization, it would be nice to have a discriminator field to use to select the type to parse into, but our API doesn't really provide that.  

Just brain dumping into this comment for further discussion.

Reply:
well, it will be a comfort python isn't the only language with a separate [de]serialization layer. I always thought it was unfair that the other languages just somehow got their models/types hydrated for free from the transport.

Reply:
## Updated feedback

- the Kotlin SDK is now using GSON to ignore serializing null values into the request payload (so far, no complaints from the few users we have of that SDK, and it resolves some use cases with some endpoints). 
- It also appears that being able to deserialize a JSON payload to a Kotlin type outside of the request/response flow is a desirable feature as well, so we may want to separate that out into an independent function